app/models/gpo_confirmation.rb

```diff
@@ -1,19 +1,65 @@
 class GpoConfirmation < ApplicationRecord
   self.table_name = 'usps_confirmations'
 
+  validates :entry, presence: true
+  validates :otp, :address1, :city, :state, :zipcode, presence: true
+  validate :zipcode_is_valid
+
+  ZIP_REGEX = /^(\d{5})[-+]?(\d+)?$/
+
+  def method_missing(method, *args)
+    # Forward through attribute reads to entry
+    entry[method] if args.count == 0
+  end
+
+  def respond_to_missing?(method, ...)
+    method = method.to_s
+    !(method.endswith('?') || method.endswith('='))
+  end
+
   # Store the pii as encrypted json
   def entry=(entry_hash)
-    self[:entry] = encryptor.encrypt(entry_hash.to_json)
+    @entry = nil
+    self[:entry] = encryptor.encrypt(
+      entry_hash.
+        dup.
+        tap do |h|
+          h[:zipcode] = self.class.normalize_zipcode(h[:zipcode]) if h[:zipcode].present?
+        end.
+        to_json,
+    )
   end
 
   # Read the pii as a decrypted hash
   def entry
-    JSON.parse(encryptor.decrypt(self[:entry]), symbolize_names: true)
+    @entry ||= JSON.parse(encryptor.decrypt(self[:entry]), symbolize_names: true)
+  end
+
+  def self.normalize_zipcode(zipcode)
+    _, zip, plus4 = ZIP_REGEX.match(zipcode&.gsub(/\s/, '')).to_a
+    if plus4&.length == 4
+      "#{zip}-#{plus4}"
+    else
+      zip
+    end
   end
```

Comment:
we have some similar code to this here, and looks like we even drop the extra 4 digits:
https://github.com/18F/identity-idp/blob/c5f9fdd923c5a3610aecc54a26f2ce12ae3ce9af/app/services/gpo_confirmation_maker.rb#L55-L61

Do you think it's worth unifying the code in these two steps of the process? Or just doing the check once earlier and having the later part of the process trust it?

Reply:
I think we should probably remove that from GpoConfirmationMaker and just let GpoConfirmation handle it. I will update